src/repositories/admin/canonical-email-block-repository.ts

```diff
@@ -0,0 +1,108 @@
+import type { MastoConfig } from '../../config';
+import { version } from '../../decorators';
+import type { Admin } from '../../entities';
+import type { Http } from '../../http';
+import { Paginator } from '../../paginator';
+import { IterableRepository } from '../iterable-repository';
+import type { DefaultPaginationParams } from '../repository';
+
+export interface TestCanonicalEmailBlockParams {
+  /** The email to canonicalize and hash */
+  readonly email: string;
+}
+
+export interface CreateCanonicalEmailBlockParamsWithEmail {
+  /** The email to canonicalize, hash, and block. If this parameter is provided, canonical_email_hash will be ignored. */
+  readonly email: string;
+  /** The hash to test against. If email is not provided, this parameter is required. */
+  readonly canonicalEmailHash?: never;
+}
+
+export interface CreateCanonicalEmailBlockParamsWithCanonicalEmailHash {
+  /** The hash to test against. If email is not provided, this parameter is required. */
+  readonly canonicalEmailHash: string;
+  /** The email to canonicalize, hash, and block. If this parameter is provided, canonical_email_hash will be ignored. */
+  readonly email?: string | null;
+}
```

Comment:
The doc says 

- email: "If this parameter is provided, canonical_email_hash will be ignored."
- canonical_email_hash: "If email is not provided, this parameter is required."

Thus, if we provide

| # | Condition | |
| :-  | :-- | :-- |
| 1 |  Only `email` | `email` will be used |
| 2 | Only `canonical_email_hash` | `canonical_email_hash` will be used |
| 3 | Neither `email` nor `canonical_email_hash` | They say "If email is not provided, this parameter is required." so it is invalid. |
| 4 | Both `email` and `canonical_email_hash` | `canonical_email_hash` will be ignored, which will work the same as ＃1 |

This in turn, the proper definition should be

```ts
export interface CreateCanonicalEmailBlockParamsWithEmail {
  /** The email to canonicalize, hash, and block. If this parameter is provided, canonical_email_hash will be ignored. */
  readonly email: string;
}

export interface CreateCanonicalEmailBlockParamsWithCanonicalEmailHash {
  /** The hash to test against. If email is not provided, this parameter is required. */
  readonly canonicalEmailHash: string;
}
```

Reply:
yh you right